FEATURE: Add Commands ``site:create``, ``site:activate`` and  ``site:decativate``

mficzel opened the pull request:
This command adds the `site:create`command that allows to create a blank site with just a single empty document in the default dimension. The name of the site, the packageKey must be specified.

If no nodeType option is specified the command will use `TYPO3.Neos.NodeTypes:Page` as fallback. The node type must already exists and have the superType `TYPO3.Neos:Document`.

If no `nodeName` option is specified the command will create a unique node-name from the name of the site. If a node name is given it has to be unique.

If the flag ``inactivate` is set the new site will not be activated.

The commands `site:activate` and `site:deactivate` are added to bring the according feature of the neos-ui to the cli.

The command `site:list` is extended to display the online state of the sites now.

Additional the  error format of the `site:import` and `site:export` command is altered to use the error format instead of an "Error:" prefix.

johannessteu commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,89 @@ class SiteCommandController extends CommandController
+     * @param boolean $active Activate the new site immediately.
+     * @return void
+     */
+    public function createCommand($name, $packageKey, $nodeName = NULL, $nodeType = 'TYPO3.Neos.NodeTypes:Page', $active = FALSE)
```
We shouldn't use `TYPO3.Neos.NodeTypes` package here as it would introduce a dependency to that package

mficzel replied:
@johannessteu i understand your point but the problem is that the only non-abstract documentNode that is part of TYPO3.Neos is TYPO3.Neos:Shortcut which is also no good default for en empty site. 

On the other hand the command would work without Neos.NodeTypes but it would require that a correct NodeType is given.

johannessteu replied:
Yeah, you're right. But i'd rather require that argument instead of using a nodeType by default wich might not be present. Even lots of ppl run commands and don't check all arguments they can use

mficzel replied:
Ok i adjusted that an altered the order of the arguments

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,89 @@ class SiteCommandController extends CommandController
+        $rootNode->getContext()->getWorkspace();
+        $sitesNode = $rootNode->getNode('/sites');
+        if ($sitesNode === null) {
+            $sitesNode = $rootNode->createSingleNode('sites');
```
`createSingleNode` is not API, please use `createNode`

mficzel replied:
Right ... is adjusted

mficzel commented:
I'm still think about moving parts of the added code to the SiteService. For the moment i came to the conclusion to do that when this feature is added to the sites management-module which is a logical next step.

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+        }
+
+        if ($this->siteRepository->findOneByNodeName($nodeName)) {
+            $this->outputLine('Error: A site with siteNodeName "%s" already exists', [$nodeName]);
```
instead of `Error:` use `<error>A site ...</error>` to output red colored text..

same with the next two..

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+     * @param boolean $active Activate the new site immediately.
+     * @return void
+     */
+    public function createCommand($name, $packageKey, $nodeType, $nodeName = null, $active = false)
```
should there be a default node type like `TYPO3.Neos.NodeTypes:Page` here? it can just thrown an error if that node type doesn't exist or is abstract and then you simply override it..

mficzel replied:
That was there but was removed to avoid a dependency to neos.nodetypes

aertmann replied:
it wouldn't be a dependency per say, I'd vote for using it as a default

johannessteu replied:
We'll it's not some kind of hard dependency but still it would only work if the Package is present. If it is not you might have some problems and a site is created you don't really can work with.

aertmann replied:
it wouldn't create a site, it would thrown an error that the node type doesn't exist.. so wouldn't cause any problems except requiring you to provide a different node type..

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+     * If no "--nodeName" is specified the command will create a unique node-name from the name of the site. If a node name is given
+     * it has to be unique.
+     *
+     * The specified "--nodeType" must it exists and have the superType "TYPO3.Neos:Document".
```
this argument comes before `--nodeName` so makes sense to move it before

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+    public function createCommand($name, $packageKey, $nodeType, $nodeName = null, $active = false)
+    {
+        if ($nodeName === null) {
+            $nodeName = $this->nodeService->generateUniqueNodeName('/sites', $name);
```
please use the constant from `SiteService`

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+     *
+     * @param string $name The name of the site
+     * @param string $packageKey The site package
+     * @param string $nodeType The nodetype to use for the site node.
```
node type

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+     * packageKey to assign to it must be specified.
+     *
+     * If no "--nodeName" is specified the command will create a unique node-name from the name of the site. If a node name is given
+     * it has to be unique.
```
might want to wrap this line a little earlier..

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+     * The specified "--nodeType" must it exists and have the superType "TYPO3.Neos:Document".
+     *
+     * If the flag "--active" is set the created site will be activated immediately.
+     *
```
should it maybe mention that the site will only be created in one dimension being the default dimension?

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+        // We fetch the workspace to be sure it's known to the persistence manager and persist all
+        // so the workspace and site node are persisted before we import any nodes to it.
+        $rootNode->getContext()->getWorkspace();
+        $sitesNode = $rootNode->getNode('/sites');
```
use constant again here

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+        $rootNode->getContext()->getWorkspace();
+        $sitesNode = $rootNode->getNode('/sites');
+        if ($sitesNode === null) {
+            $sitesNode = $rootNode->createNode('sites');
```
and here

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,88 @@ class SiteCommandController extends CommandController
+        if ($siteNodeType->isOfType('TYPO3.Neos:Document') === false) {
+            $this->outputLine('Error: the given root nodeType "%s" is not based on the superType "%s"', [$nodeType, 'TYPO3.Neos:Document']);
+            $this->quit(1);
+        }
```
I'd suggest to move this up before starting to create the root site node

aertmann commented:
Hey Martin very nice addition! Left some comments for improvement.

mficzel commented:
@aertmann thanks for the nitpicks i will update the pr soon. 

Regarding the default nodeType i have a suggestion: 
We could add a TYPO3.Neos.SiteNodeTypes Section to Settings.yaml where the available RootNodeTypes are defined. That way 'TYPO3.Neos.NodeTypes:Page' could be added to that list from the NodeTypes Package itself. The advantage would be that we can use this list to make those nodeTypes available in the ui of the sites-module/create action. Does this make sense to you?

kitsunet commented:
Which nodes could be used for the site node should be easily derived from the list of document nodes and eventually constraints. No need for a configuration there. Also I think it's fine to just require this argument.

aertmann commented:
sounds like overcomplicating things, for the UI just allow all document node types which aren't constrained on that path like Christian suggests.. like I said, there's no harm in providing a default that can be overridden if desired or the package isn't present

mficzel commented:
I think i adjusted the mentioned points. So the next round of feedback is welcome.

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,94 @@ class SiteCommandController extends CommandController
+
+        $rootNode = $this->nodeContextFactory->create()->getRootNode();
+        // We fetch the workspace to be sure it's known to the persistence manager and persist all
+        // so the workspace and site node are persisted before we import any nodes to it.
```
Mmm, I am not seeing any persist call, also we are talking about the live workspace. It should exist already. I would skip all that and do a getNode() with the SITES_ROOT_PATH directly on the context created by the factory.

mficzel replied:
I am not sure that the sites-node always exists. What about a fresh setup via cli. 

About persisting. Why should i do that explicit flow does that already?

kitsunet replied:
I just mean the comment says it is persisted before we import any nodes to it. Seems misleading. I expected a persistCall before creating the node(s).
So maybe just remove the comment?

mficzel replied:
Ahh ... now i got it it is about the following `$rootNode->getContext()->getWorkspace();` copied that  from the SiteImportService service and i forgot the `$this->persistenceManager->persistAll();`. I am not entirely shure why this has to be done that way but since the siteImportService does it i simply do the same here.

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,94 @@ class SiteCommandController extends CommandController
+
+        $siteNode = $sitesNode->createNode($nodeName, $siteNodeType);
+        $siteNode->setProperty('title', $name);
+
```
NodePaths::getNodeNameFromPath() pls

kitsunet commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -63,6 +67,94 @@ class SiteCommandController extends CommandController
+        $rootNode->getContext()->getWorkspace();
+        $sitesNode = $rootNode->getNode(SiteService::SITES_ROOT_PATH);
+        if ($sitesNode === null) {
+            $sitesNode = $rootNode->createNode(trim(SiteService::SITES_ROOT_PATH, '/'));
```
NodePaths::getNodeNameFromPath() pls

kitsunet commented:
Two smallish comments, could also be refactored later. Good work 👍

mficzel commented:
I pushed an update ... both things you mentioned resulted in a lazy copy of the same sequence in the siteImportService. Maybe i looked at an older version there. We should think about extracting that to the SiteService in future.

robertlemke commented:
Thanks for this new command!

I didn't read all the code, so here comes a question solely from a user's perspective. I tried the following:

- `./flow site:prune`
- `./flow site:create "Neos Demo Site" Neos.Demo

and then hit the front page in the browser. That shows me a "homepage not found".
I also flushed caches to check if the not-found was due to the routing or content cache (by the way, should that be tackled in the `site:create` command?) but that didn't help either.

Just to see if my setup is generally fine, I did a `./flow site:import Neos.NeosDemo` again and verified that the homepage shows up.

mficzel commented:
@robertlemke thanks for testing ... the site that is created is not active by default. The complete cli call to directly activate the site would be `./flow site:create "Neos Demo Site" Neos.Demo --active` that gives you single empty page  with an ts-error in the footer because the footer is fetched from another document.

mficzel commented:
@robertlemke i just noticed that the current cli has no options to determine wether a site is online or not. I added this and also the site:activate and site:deactivate commands.

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -213,7 +314,8 @@ public function listCommand()
                 'nodeName' => $site->getNodeName(),
-                'siteResourcesPackageKey' => $site->getSiteResourcesPackageKey()
+                'siteResourcesPackageKey' => $site->getSiteResourcesPackageKey(),
+                'status' => ($site->getState() == SITE::STATE_ONLINE) ? 'online' : 'offline'
```
===

robertlemke commented:
@mficzel what would be the drawback of activating a site right after its creation? I'd argue that most people use that command during development and not on live websites. And it's a little WTF moment when you create a site and it's not there (see my dummy-user comments earlier).

How about making it active by default and providing an option `activate` which is true by default? (instead of the `active` option).

mficzel commented:
@robert ... i renamed ``active`` to ``activate`` and made it true by default as you suggested. To deactivate a site after creation you now would type ``./flow site:create --activate 0  "Neos Demo" Neos.Demo``

mficzel commented:
The use case i have in mind is a multi site setup with multiple sites that share a site package. So I would use this in production.

mficzel commented:
Rebased and renamed the ``active`` option to `ìnactive`` to make the most likely use case the default.

So:
- ``./flow site:create "Neos Demo Site" Neos.Demo`` will create and activate the site 
- ``./flow site:create "Neos Demo Site" Neos.Demo --inactive``  will just create the site without activation

mficzel commented:
A related pr that adds the same feature to the admin-module is here #757

mficzel commented:
@robertlemke can you check again or at least remove the "thumbs down" i adjusted the point you mentioned a while ago.

robertlemke commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -65,6 +70,101 @@ class SiteCommandController extends CommandController
+
+    /**
+     * Create a new (blank) site in the default dimension
+     *
```
Nitpick: "Create a new (blank) site in the default dimension" is a bit long for the help screen, I'll change it to something shorter

robertlemke commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -229,11 +330,48 @@ public function listCommand()
+     * @param string $siteNodeName The node name of the site to activate
+     * @return void
+     */
+    public function activateCommand($siteNodeName)
```
Please let's stay consistent with the command argument name here, like in `site:export`: even though `$siteNodeName` is more correct from the PHP side, we once decided that `--site-node` is better for the CLI side.

`site:prune` unfortunately defers from that rule already, but that's another story.

robertlemke commented on TYPO3.Neos/Classes/TYPO3/Neos/Command/SiteCommandController.php:
```diff
@@ -229,11 +330,48 @@ public function listCommand()
+     * @return void
+     */
+    public function deactivateCommand($siteNodeName)
+    {
```
Same like in `activateCommand`.

robertlemke commented:
@mficzel I left a few comments regarding the naming and took the liberty to adjust them right away in a new commit. Please speak up if you don't agree.

I'm ashamed to say that I didn't try this patch for real 😊. The changes I committed were written directly on Github. So, if you can do me a favour to quickly give the patch a final test on the command line? If it works as expected, feel free to merge!
